File: Data Structures & Algorithms/max-water-container/submission-0.java
// 1,7,2,5,4,7,3,6
//   l
//     r

class Solution {
    public int maxArea(int[] heights) {
        int maxArea = 0;
        int left = 0;
        int right = heights.length - 1;

        while(left < right) {
            int width = right - left;
            int height = Math.min(heights[left], heights[right]);

            maxArea = Math.max(maxArea, width * height);

            if(heights[left] < heights[right]) {
                left++;
            }
            else {
                right--;
            }
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        max_val = 0

        def traverse(root):
            nonlocal max_val
            #Base case
            if root is None:
                return 0

            dl = traverse(root.left)
            dr = traverse(root.right)

            max_val = max(dl + dr , max_val)

            return max(dl, dr) + 1

        
        traverse(root)
        return max_val

        
        